docs/migrations/chain-upgrade-guide-040.md

```diff
@@ -0,0 +1,228 @@
+* Prior to exporting the state, the validators are encouraged to take a full data snapshot at exported height. Exported
+height will be determined by a governance proposal. Data backup is usually done by copying daemon home directory,
+ex: `~/.gaiad`
+
+It is critically important to back-up the validator state file, ex: `~/.gaiad/data/priv_validator_state.json` file
+after stopping your daemon process. This file is updated every block as your validator participates in a consensus
+rounds. It is a critical file needed to prevent double-signing, in case the upgrade fails and the previous chain needs
+to be restarted.
+
+In the event that the upgrade does not succeed, validators and operators must downgrade back to old version of the
+software and restore to their latest snapshot before restarting their nodes.
+
+### Upgrade procedure
+- Use old binary to export the state. Make sure to verify your binary version before exporting the state
+- Export the state from existing chain using old-binary (which uses `sdk@0.39.x`).
+Example:
+    ```sh
+    gaiad export --for-zero-height --height <height> > 039_exported_state.json
+    ```
+- Verify the SHA256 of the (sorted) exported genesis file:
+    ```shell
+    $ jq -S -c -M '' 039_exported_state.json | shasum -a 256
+    [PLACEHOLDER]  039_exported_state.json
+    ```
+- Cross check the hash with other peers (other validators) in the chat rooms
+- Install the latest binary (which uses `0.40`)
+- Migrate the exported state to `0.40` compatible genesis state
+    ```shell
+    gaiad migrate v40 039_exported_state.json --chain-id <new_chain_id> --genesis-time <new_genesis_time_in_utc> > new_v40_genesis.json
+    ```
+  NOTE: The migrate command takes an input genesis state and migrates it to a targeted version. New `genesis-time` will
+  be as mentioned in the governance proposal.
+- Verify the SHA256 of the migrated genesis file with other valdiators to make sure there are no manual errors in the process.
+    ```shell
+    $ jq -S -c -M '' new_v40_genesis.json | shasum -a 256
+    [PLACEHOLDER]  new_v40_genesis.json
+    ```
+- Make sure to update the genesis parameters in the new genesis if any. All these details will be generally present in
+the governance proposal
+- All the necessary state chanegs are handled in `040` migration command, including tendermint params. So, manual updates to
+the genesis are not required
+- Reset the old state
+NOTE: Be sure you have a complete backed up state of your node before proceeding with this step.
+See Recovery for details on how to proceed.
+```shell
+gaiad unsafe-reset-all
+```
+- Move the new genesis.json to your daemon config directory. Ex
+```shell script
+cp new_v40_genesis.json ~/.gaiad/config/genesis.json
+```
+
+- Update `~/.gaiad/config/app.toml` to include latest app configurations. Below is an example `app.toml`. Make sure to
+update your custom configurations as per your validator design ex: `gas_price`.
+
+```yaml
+# This is a TOML config file.
```

Comment:
I don't think it's necessary to paste the toml file here.

Reply:
There are a number of updates required to existing app.toml, so if someone is upgrading the chain from 039 to 040, this file would be handy. May be it can go into a different place and we can link it here. wdyt?

Reply:
Ideally, this document would only show the fields that need change, e.g. in a markdown table.

If there are too many changes, we could put the toml in a separate file, and use `+++` in markdown to paste it.